pkg/compactor/bucket_compactor.go

```diff
@@ -422,6 +429,59 @@ func (c *BucketCompactor) runCompactionJob(ctx context.Context, job *Job) (shoul
 	return true, compIDs, nil
 }
 
+// verifyCompactedBlocksTimeRanges does a full run over the compacted output blocks
+// and verifies that they satisfy the min/maxTime from the input blocks
+func verifyCompactedBlocksTimeRanges(ctx context.Context, compIDs []ulid.ULID, inputBlocksMinTime, inputBlocksMaxTime int64, blockSyncConcurrency int, subDir string) error {
+	inputBlocksMinTimeFound := atomic.NewBool(false)
+	inputBlocksMaxTimeFound := atomic.NewBool(false)
+
+	err := concurrency.ForEachJob(ctx, len(compIDs), blockSyncConcurrency, func(ctx context.Context, idx int) error {
+		outputBlock := compIDs[idx]
+
+		// Skip empty block
+		if outputBlock == (ulid.ULID{}) {
+			return nil
+		}
+
+		bdir := filepath.Join(subDir, outputBlock.String())
+		meta, err := block.ReadMetaFromDir(bdir)
+		if err != nil {
+			return errors.Wrapf(err, "failed to read meta.json from %s", bdir)
+		}
+
+		// Ensure meta min/maxTime within input min/maxTime
+		if meta.MinTime < inputBlocksMinTime {
+			return fmt.Errorf("invalid minTime for block %s, block minTime %d is before input minTime %d", outputBlock.String(), meta.MinTime, inputBlocksMinTime)
```

Comment:
[nit] We're used to call the input blocks as "source blocks" and output blocks as "compacted blocks". I recommend to keep naming consistent in the errors and variables too.